### PR TITLE
IGNITE-17167 ConfigurationAsmGenerator decomposition

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-17167

### CODE REVIEW

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/Methods.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+import org.apache.ignite.internal.configuration.tree.ConfigurationVisitor;
+import org.apache.ignite.internal.configuration.tree.ConstructableTreeNode;
+import org.apache.ignite.internal.configuration.tree.InnerNode;
+import org.apache.ignite.internal.configuration.tree.NamedListNode;
+import org.apache.ignite.internal.configuration.util.ConfigurationUtil;
+
+/**
+ * Class that holds {@link Method} constants to be used by generators.
+ */
+public class Methods {
```
Comment (tkalkirill) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/Methods.java:46]: Why `public` ? I see that there is inheritance from it, it seems it should be `abstract`.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/CfgImplAsmGenerator.java [hunk 2]
```diff
@@ -0,0 +1,900 @@
...
+import org.apache.ignite.internal.configuration.DynamicConfigurationChanger;
+import org.apache.ignite.internal.configuration.DynamicProperty;
+import org.apache.ignite.internal.configuration.NamedListConfiguration;
+import org.apache.ignite.internal.configuration.direct.DirectPropertyProxy;
+import org.apache.ignite.internal.configuration.tree.InnerNode;
+import org.jetbrains.annotations.Nullable;
+import org.objectweb.asm.Handle;
+import org.objectweb.asm.Opcodes;
+
+class CfgImplAsmGenerator extends Methods {
```
Comment (tkalkirill) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/CfgImplAsmGenerator.java:98]: Why such an abbreviation for the class? `ConfigImplAsmGenerator` or `ConfigurationImplAsmGenerator` better.
This is not a complete rework, but only a "combing", please indicate a ticket in which we will continue to improve the code.

**Review by tkalkirill — APPROVED** [review]

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/Methods.java (hunk 1 above)
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/Methods.java]: Made abstract and package-private, thank you!

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/CfgImplAsmGenerator.java (hunk 2 above)
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/CfgImplAsmGenerator.java]: No specific reasons, I renamed this class

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/InnerNodeAsmGenerator.java [hunk 3]
```diff
@@ -0,0 +1,1787 @@
...
+                        constantClass(polymorphicExtension).ret()
+                );
+            }
+
+            mtdBody.append(switchBuilderTypeId.build());
+        }
+    }
+
+    /**
+     * Declares field that corresponds to configuration value. Depending on the schema, 5 options possible:
```
Comment (rpuch) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/InnerNodeAsmGenerator.java:363]:
```suggestion
     * Declares a field that corresponds to configuration value. Depending on the schema, 5 options are possible:
```

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java [hunk 4]
```diff
@@ -570,1669 +312,116 @@ private Field internalIdField(Class<?> schemaClass, Set<Class<?>> schemaExtensio
...
-
-        return classDef.declareField(of(PUBLIC), fieldName, nodeFieldType);
+    private static String capitalize(String name) {
+        return name.substring(0, 1).toUpperCase() + name.substring(1);
     }
 
     /**
-     * Implements default constructor for the node class. It initializes {@code _spec} field and every other field that represents named
-     * list configuration.
+     * Returns internalized version of class name, replacing dots with slashes.
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java:385]: "internal name" is a thing in JVM, I didn't choose it. I'll chance the sentence a little bit anyway to make it simpler to understand.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java [hunk 5]
```diff
@@ -570,1669 +312,116 @@ private Field internalIdField(Class<?> schemaClass, Set<Class<?>> schemaExtensio
...
-
-            // this._internalConfigTypes = tmp;
-            initInternalSchemaTypesField.append(setThisFieldCode(ctor, tmpVar, internalSchemaTypesFieldDef));
-
-            ctorBody.append(initInternalSchemaTypesField);
-        }
-
-        // return;
-        ctorBody.ret();
+    @NotNull
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java:391]: Sorry, my bad

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationImplAsmGenerator.java [hunk 6]
```diff
@@ -0,0 +1,872 @@
...
+
+        methodDef.getBody().retObject();
+    }
+
+    /**
+     * Implements accessor method in configuration impl class.
+     *
+     * @param classDef    Configuration impl class definition.
+     * @param schemaField Configuration Schema class field.
+     * @param fieldDefs   Field definitions.
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationImplAsmGenerator.java:492]: I wouldn't want to change the code that I moved too much :) Bu I'll do it here. Just wanted to remind that I didn't write this code.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java [hunk 7]
```diff
@@ -3251,7 +486,7 @@ private static StringSwitchBuilder typeIdSwitchBuilder(MethodDefinition mtdDef,
      * @param fieldDefs Field definitions.
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java:486]: It is already mentioned in the description section of the method, I wouldn't want to repeat myself.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/AbstractAsmGenerator.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    static final String CONSTRUCT_MTD_NAME = "construct";
+
+    /** Field name for method {@link DynamicConfiguration#internalConfigTypes}. */
+    static final String INTERNAL_CONFIG_TYPES_FIELD_NAME = "_internalConfigTypes";
+
+    /** {@link DirectConfigurationProxy#DirectConfigurationProxy(List, DynamicConfigurationChanger)}. */
+    static final Constructor<?> DIRECT_CFG_CTOR;
+
+    /** {@link ConfigurationUtil#appendKey(List, Object)}. */
+    static final Method APPEND_KEY;
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/AbstractAsmGenerator.java:147]: That's a laziness, I hated these constants in every class

**Pull request merged**